### ISSUE TITLE
CFP: Allow certain pods to emit traffic from other IPs

### ISSUE THREAD

**Issue by or-adar** [cilium/cilium#22281]
Hi All,
I'm using Ciliun CNI and it drops packets with the message "Invalid  source ip".
In my cluster, I have a pod running a proxy service (envoy) which forwards packets of various clients to a Pod running a VPN server.
I have to preserve the source IPs of my clients, so when the envoy Pod forwards packets to the desired Pod - the src IP is still the address of the client which initiated the request (and not the IP of my envoy Pod).

to sum it up, the packet flow is: `client -> envoy Pod -> VPN server Pod`
and the src ip remains the same for every hop (and determined by the IP of the client)
so as a result the envoy Pod emits traffic with src IP that is not it's own IP.

I think I understand the reason why it's being blocked by cilium,
so what I want to understand is how can I disable the behaviour to allow my pod to emit traffic with packets containg other src IPs?

**Comment by joestringer** [comment]
@or-adar and I discussed this briefly on Slack the other day, so I'll summarize here: This feature is enabled in Cilium by default to prevent pods from spoofing the source IP address of other networking elements in the cluster. As of 1.13.x development versions, there is no way to turn this off in the codebase. For some context on the sort of security issues that can come up without this protection, see https://cilium.io/blog/2019/07/01/cve-2019-13119/ .

Some other community members are looking into enabling some VPN use cases with more concrete route configuration as Cilium CRDs, see this proposal: https://docs.google.com/document/d/1Sjabt_z6CRIyDgMk8ZbhiXTaiK5RnR982AO4h6ZYLvI/edit#heading=h.5mwz58y9syps (I don't think this has a corresponding issue right now).

One of the questions I asked was, let's say that this traffic wasn't blocked by Cilium. How do you make sure that the replies go back into Envoy if the source IP addresses are somewhat random? This would be important for terminated TCP connections, because Envoy is not guaranteed to match the TCP windowing of the original connection. Returning traffic directly to the original source will likely result in the packets being dropped / ignored and breaking the connection.

The relevant code in the datapath stems from usage of the `LXC_IP` variable if you want to pull on a thread to investigate changes yourself for this sort of feature. Given the security implications, I am not sure whether it would make sense to accept a change to this behaviour into Cilium unless there is a well considered security review of the change.

**Comment by or-adar** [comment]
Yeah, sorry for spamming I posted both questions on slack & here 
as I didn't know what platform would be more relevant to discuss about it.

TCP traffic is potentially problematic as you stated, but I'm more focused on UDP-based traffic as of now.

My use case is to use envoy as a replacement for LoadBalancer to ultimately save money,
Similarly to how it was presented [here](https://github.com/cilium/cilium/issues/22281) but using ClusterIP instead of NodePort.
Also, a side note: I tested a setup with Cilium using L4 LoadBalancer with `service.spec.externalTrafficPolicy` set to `Local` (persisting src ip) and that setup worked fine without issues.
So maybe a follow up question: how come Cilium do not drop packets when using LB and persisting src ip?
Why is there a difference when having a proxy service (envoy) exposed to the internet to forward traffic to internal Pods (while persisting src ips)?

Appreciate your help and tips! and I'll be looking at the usage of the `LXC_IP` variable.

**Comment by joestringer** [comment]
The key difference between Slack and here is that Slack is more ephemeral. It can be helpful for more active synchronous discussions and brainstorming, but the history will be lost within a few months.

> how come Cilium do not drop packets when using LB and persisting src ip?

This is quite an insightful question actually. Typically Cilium assumes that applications running in the pods are untrusted, and that an adversary may have control over the code executed in that environment. As such, if someone is able to cause that container to spoof traffic then Cilium should not allow that to then escalate to wider Cilium node or cluster privileges/access. On the other hand, when implementing LB the traffic is presumably either coming into the environment from outside, in which case it is likely that there are some protections around spoofing and the source of that traffic, or alternatively someone has access to the underlying network, at which point they could generate that traffic regardless of Cilium's role in LB so Cilium does not provide a vector for privilege escalation in the same sense.

**Comment by doup123** [comment]
I am trying to deploy an application that requires to be able to spoof packets (https://github.com/sleinen/samplicator); it actually replicates traffic to multiple destinations by keeping unchanged the original IP address.
Did you @or-adar @joestringer by any chance find a way to do this?
